### PR TITLE
Configurable telemetry event filtering

### PR DESCRIPTION
**Pull request by mlw**
Adds support for a new `Telemetry` configuration key that allows admins to specify the event types that should be logged.

This PR deprecated the `EnableForkAndExitLogging` key.

### CODE REVIEW

File: Source/common/SNTConfigurator.m [hunk 1]
```diff
@@ -1285,9 +1304,9 @@ - (void)applyOverrides:(NSMutableDictionary *)forcedConfig {
   NSDictionary *overrides = [NSDictionary dictionaryWithContentsOfFile:kConfigOverrideFilePath];
   for (NSString *key in overrides) {
     id obj = overrides[key];
-    if (![obj isKindOfClass:self.forcedConfigKeyTypes[key]] ||
-        ([self.forcedConfigKeyTypes[key] isKindOfClass:[NSRegularExpression class]] &&
-         ![obj isKindOfClass:[NSString class]])) {
+    if (![obj isKindOfClass:self.forcedConfigKeyTypes[key]] &&
+        !(self.forcedConfigKeyTypes[key] == [NSRegularExpression class] &&
+          [obj isKindOfClass:[NSString class]])) {
```
Comment (mlw) [inline Source/common/SNTConfigurator.m:1309]: Note: This is an unrelated fix to being unable to set `re` type keys in the config overrides file.

**Review by pmarkowsky — CHANGES_REQUESTED** [review]
Looks good. Some minor things.

File: Source/santad/EventProviders/SNTEndpointSecurityRecorder.mm [hunk 2]
```diff
@@ -130,6 +131,13 @@ - (void)handleMessage:(Message &&)esMsg
 
   [self.compilerController handleEvent:esMsg withLogger:self->_logger];
 
+  // The logger will take care of this, but we check early so we
+  // don't do any unnecessary work
+  if (!_logger->ShouldLog(santa::ESEventToTelemetryEvent(esMsg->event_type))) {
+    recordEventMetrics(EventDisposition::kDropped);
```
[the remark this answers is not in the document]
Comment (mlw) [reply Source/santad/EventProviders/SNTEndpointSecurityRecorder.mm:137]: I went back and forth on whether metrics should even be recorded... Ideally we never get these events in the first place (e.g. we drop subscriptions), however that's only partially true due to some event types we must always subscribe to.

The old `EnableForkAndExitLogging` key resulted in `EventDisposition::kDropped` being recorded so I opted to keep that here for all event types.

As for whether or not it should be a new disposition - maybe? I think there are a couple of inconsistent places across all ES clients for how kProcessed vs. kDropped are used. The original intention was basically to say "this many were logged" and "this many were received but not logged".  This is a bit simplistic though for how complicated the clients are... Would be nice to put some thought into this for the long term for how we'd want these represented.

[the remark this answers is not in the document]
Comment (pmarkowsky) [reply Source/santad/EventProviders/SNTEndpointSecurityRecorder.mm:137]: Given that this isn't consistent we should solve that in a future PR. I've opened https://github.com/northpolesec/santa/issues/150 to track it.

File: Source/santad/Santad.mm [hunk 3]
```diff
@@ -398,6 +400,47 @@ void SantadMain(std::shared_ptr<EndpointSecurityAPI> esapi, std::shared_ptr<Logg
                                             FlushCacheReason::kEntitlementsPrefixFilterChanged);
               [authorizer_client clearCache];
             }],
+    [[SNTKVOManager alloc] initWithObject:configurator
+                                 selector:@selector(telemetry)
+                                     type:[NSArray class]
+                                 callback:^(NSArray *oldValue, NSArray *newValue) {
+                                   if (!oldValue && !newValue) {
+                                     return;
```
[the remark this answers is not in the document]
Comment (mlw) [reply Source/santad/Santad.mm:408]: This condition isn't quite an error case, just protects against the case where we sometimes get KVO callbacks when things don't actually change.

File: Source/common/SNTConfigurator.m [hunk 4]
```diff
@@ -1038,6 +1045,18 @@ - (BOOL)enableForkAndExitLogging {
...
 }
 
+- (NSArray<NSString *> *)telemetry {
+  NSArray *events = self.configState[kTelemetryKey];
+
+  for (id event in events) {
+    if (![event isKindOfClass:[NSString class]]) {
+      return nil;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mlw) [reply Source/common/SNTConfigurator.m]: I changed this up a bit after thinking about it... If someone attempts to filter, "failing open" and allowing all events seems wrong. Instead, it will now strip out non-string things in a sort of "fail closed" operation. Added a comment to explain.

**Review by pmarkowsky — APPROVED** [review]
LGTM.
